File: docs/codes/monitoring/monitoring.py
"""
Usage:
     DataRobot's Monitoring Agent is an advanced feature that enables monitoring for models deployed
     outside of DataRobot. Follow these steps to use this feature.
    1. Download and extract DataRobot's Monitoring Agent tar file. This is available through the
       DataRobot MLOps UI via User icon -> "Developer Tools" -> "External Monitoring Agent".
    2. Open your preferred browser. In toolbar, click "File" -> "Open File".
       Then choose this file <your unpacked directory>/docs/html/quickstart.html.
    3. Follow the "Quick Start" instructions to set up Monitoring Agent.
    4. This example uses DataRobot's MLOps Python library which you can install with:
       pip install <unpacked DataRobot's Monitoring Agent tar file>/lib/datarobot_mlops-*-py2.py3-none-any.whl
    5. The MLOps library requires the following parameters to be provided:
       deployment ID, model ID, and spooler type [FILESYSTEM, SQS, RABBITMQ, PUBSUB, STDOUT, NONE].
       The spooler type selected may require additional configuration. For example,
       if the spooler type is FILESYSTEM, the filesystem directory will have to be configured.
       The spooler directory path must match the Monitoring Agent path configured by the administrator.
       For more details and advanced usage, see the examples and documentation included with the agent tar file.
       These parameters can be configured with the MLOps library APIs or with environment variables.
       For example:
       # export MLOPS_DEPLOYMENT_ID=60f0443ed7733e8161155078
       # export MLOPS_MODEL_ID=60f03f8e0a3504f0d3764a06
       # export MLOPS_SPOOLER_TYPE=FILESYSTEM
       # export MLOPS_FILESYSTEM_DIRECTORY=/tmp/ta
 
       Notes:
              - parameter configuration via environment variables takes precedence over the API.
              - for testing purposes, you can start with STDOUT output type,
                which doesn't require providing any of the spooler-related parameters. The reporting results
                are sent to stdout and not forwarded to the Monitoring Agent or DataRobot MLOps.
              - complete examples are included in the agent tar file in the `examples` directory.
    6. Run current snippet:
       python datarobot-report-stats.py
"""
 
import sys
import time
import random
import pandas as pd
 
from datarobot.mlops.mlops import MLOps
from datarobot.mlops.common.enums import OutputType
 
DEPLOYMENT_ID = '60f0443ed7733e8161155078'
MODEL_ID = '60f03f8e0a3504f0d3764a06'
# Spool directory path must match the Monitoring Agent path configured by admin.
SPOOL_DIR = "/tmp/ta"
 
"""
This sample code demonstrates usage of the MLOps library.
It does not have real data (or even a real model) and should not be run against a real MLOps
service.
"""
 
 
class FakeModel(object):
    """
    Note that this FakeModel is not real.
    Update it with your own model.
    Fill in the predictions and the feature details.
    """
    def predict(self, num_samples):
        """
        This function returns a list of predictions.
 
        Fill in the prediction data.
        """
        return [random.random() for i in range(num_samples)]
 
    def feature_list(self, num_samples):
        """
        This function returns feature data in a dataframe.
        Typically, the scoring data will be read in a dataframe from the CSV file
 
        Fill in the feature data.
        """
        f1 = [random.random() for i in range(num_samples)]
        f2 = [random.random() for i in range(num_samples)]
        f3 = [random.random() for i in range(num_samples)]
 
        # create a dictionary of feature name to array of feature values
        feature_data = {
            "feature1": f1,
            "feature2": f2,
            "feature3": f3
        }
 
        return pd.DataFrame.from_dict(feature_data)
 
 
def main(deployment_id, model_id, spool_dir):
    """
    This is a regression algorithm example.
    User can call the DataRobot MLOps library functions to report statistics.
    """
    # MLOPS: initialize the MLOps instance
    mlops = MLOps() \
        .set_deployment_id(deployment_id) \
        .set_model_id(model_id) \
        .set_filesystem_spooler(spool_dir) \
        .init()
 
    num_samples = 10
    # Create fake model instance
    fake_model = FakeModel()
    features_df = fake_model.feature_list(num_samples)
 
    # Get predictions
    start_time = time.time()
    predictions = fake_model.predict(num_samples)
    num_predictions = len(predictions)
    end_time = time.time()
 
    # MLOPS: report the number of predictions in the request and the execution time.
    mlops.report_deployment_stats(num_predictions, end_time - start_time)
 
    
    # MLOPS: report the predictions data: features, predictions
    mlops.report_predictions_data(features_df=features_df, predictions=predictions)
    
 
    # MLOPS: release MLOps resources when finished.
    mlops.shutdown()
 
 
if __name__ == "__main__":
    sys.exit(main(DEPLOYMENT_ID, MODEL_ID, SPOOL_DIR))
 

File: docs/codes/monitoring/monitoring.java
/**
 *  Usage:
 *  DataRobot's Monitoring Agent is an advanced feature that enables monitoring for models deployed
 *  outside of DataRobot. Follow these steps to use this feature.
 *  1. Download and extract DataRobot's Monitoring Agent tar file. This is available through the
 *     DataRobot MLOps UI via User icon -> "Developer Tools" -> "External Monitoring Agent".
 *  2. Open your preferred browser. In toolbar, click "File" -> "Open File".
 *     Then choose this file <your unpacked directory>/docs/html/quickstart.html.
 *  3. Follow the "Quick Start" instructions to set up the Monitoring Agent.
 *  4. Install the datarobot-mlops jar file into the local mvn repository.
 *     Go to <your unpacked directory>/examples/java directory, run:
 *     ./install_jar_into_maven.sh
 *  5. Include datarobot-mlops maven dependency in your pom.xml
 *     <dependency>
 *         <groupId>com.datarobot</groupId>
 *         <artifactId>datarobot-mlops</artifactId>
 *         <version>[6.2,)</version>
 *     </dependency>
 *  6. Compile your java code:
 *     mvn package
 *  7. The MLOps library requires the following parameters to be provided:
 *     deployment ID, model ID, and spooler type [FILESYSTEM, SQS, RABBITMQ, PUBSUB, STDOUT, NONE].
 *     The spooler type selected may require additional configuration. For example,
 *     if the spooler type is FILESYSTEM, the filesystem directory will have to be configured.
 *     The spooler directory path must match the Monitoring Agent path configured by the administrator.
 *     For more details and advanced usage, see the examples and documentation included with the agent tar file.
 *     These parameters can be configured with the MLOps library APIs or with environment variables.
 *     For example:
 *     # export MLOPS_DEPLOYMENT_ID=60f0443ed7733e8161155078
 *     # export MLOPS_MODEL_ID=60f03f8e0a3504f0d3764a06
 *     # export MLOPS_SPOOLER_TYPE=OUTPUT_DIR
 *     # export MLOPS_FILESYSTEM_DIRECTORY=/tmp/ta
 *  8. Execute the java example:
 *     java -jar <jar_file_created_in_step_6> <datarobot_model_jar_file> <dataset_in_csv_format>
 *
 *     Notes:
 *            - parameter configuration via environment variables takes precedence over the API.
 *            - for testing purposes, you can start with STDOUT spooler type,
 *              which doesn't require providing any of the spooler-related parameters. The reporting results
 *              are sent to stdout and not forwarded to the Monitoring Agent or DataRobot MLOps.
 *            - depending on how your Java package is defined, provide a package directive for this file, e.g:
 *              package com.datarobot.mlops.examples;
 *            - complete examples are included in the agent tar file in the `examples` directory.
 */
 
import com.datarobot.mlops.MLOps;
import com.datarobot.mlops.common.exceptions.DRCommonException;
import com.datarobot.prediction.IPredictorInfo;
import com.datarobot.prediction.IRegressionPredictor;
import com.datarobot.prediction.Predictors;
import com.opencsv.CSVReader;
import com.opencsv.CSVReaderBuilder;
 
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.net.URL;
import java.net.URLClassLoader;
 
public class MLOpsCodeGenRegressionExample {
    private String modelFilePath;
    private String csvFilePath;
    private String[] sampleHeaders;
    private IRegressionPredictor predictor;
 
    MLOpsCodeGenRegressionExample(String modelFilePath, String csvFilePath) {
        this.modelFilePath = modelFilePath;
        this.csvFilePath = csvFilePath;
    }
 
    public List<Map<String, Object>> readAllDataAtOnce(String file) throws IOException {
        List<String[]> allData = null;
        try {
            // Create an object of file reader
            // class with CSV file as a parameter.
            FileReader filereader = new FileReader(file);
 
            // create csvReader object
            CSVReader csvReader = new CSVReaderBuilder(filereader)
                    .build();
            allData = csvReader.readAll();
        } catch (IOException e) {
            System.err.println("Failed to read data file - " + file);
            throw e;
        }
 
        List<Map<String, Object>> listOfSamples = new ArrayList<>();
        this.sampleHeaders = allData.get(0);
 
        for (String[] row : allData.subList(1, allData.size())) {
            Map<String, Object> sample = new HashMap<>();
            for (int i = 0; i < this.sampleHeaders.length; ++i) {
                sample.put(this.sampleHeaders[i], row[i]);
            }
            listOfSamples.add(sample);
        }
        return listOfSamples;
    }
 
    /**
     * Load the CodeGen model from the model jar file
     */
    void loadModel() throws IOException, DRCommonException {
        try {
            File filePath = new File(modelFilePath);
            URL[] urls = new URL[]{new URL("file://" + filePath.getCanonicalPath())};
            URLClassLoader urlClassLoader = new URLClassLoader(urls);
            IPredictorInfo predictorInfo = Predictors.getPredictor(urlClassLoader);
 
            if (!(predictorInfo instanceof IRegressionPredictor)) {
                throw new DRCommonException("Provided model is not a Java CodeGen regression model");
            }
 
            this.predictor = (IRegressionPredictor) predictorInfo;
 
        } catch (IOException e) {
            System.err.println("Failed to load model. " + e.getMessage());
            throw e;
        }
    }
 
    /**
     * Call the CodeGen model to make predictions on the samples
     * @param samples
     * @return
     */
    List<Double> makePredictions(List<Map<String, Object>> samples) {
        List<Double> predictions = new ArrayList<>();
 
        for (Map<String, Object> sample : samples) {
            // return map of class label to its probability
            Double score = predictor.score(sample);
 
            predictions.add(score);
        }
 
        return predictions;
    }
 
 
    HashMap<String, List<Object>> getFeaturesFromSamples(List<Map<String, Object>> samples) {
        HashMap<String, List<Object>> featureData = new HashMap<>();
 
        for (Map<String, Object> sample : samples) {
            for (Map.Entry<String, Object> entry : sample.entrySet()) {
                List<Object> featureValues = featureData.get(entry.getKey());
                if (featureValues == null) {
                    featureValues = new ArrayList<>();
                    featureData.put(entry.getKey(), featureValues);
                }
                featureValues.add(entry.getValue());
            }
        }
 
        return featureData;
    }
 
    /**
     * Sample code demonstrating how to use MLOps library to record metrics on classification predictions.
     * All code specific to MLOps library are commented with "MLOPS"
     */
    public void run() {
        // MLOPS: declaration
        MLOps mlops = null;
        try {
            loadModel();
 
            List<Map<String, Object>> samples = readAllDataAtOnce(this.csvFilePath);
 
            HashMap<String, List<Object>> featureData = getFeaturesFromSamples(samples);
 
            // MLOPS: initialize mlops instance
            mlops = MLOps.getInstance()
                    .setDeploymentId("60f0443ed7733e8161155078")
                    .setModelId("60f03f8e0a3504f0d3764a06")
                    .setFilesystemSpooler("/tmp/ta")
                    .init();
 
            // Call the model to generate predictions.
            // For this example, we simply fake the predictions.
            long startTime = System.currentTimeMillis();
            List<Double> predictions = makePredictions(samples);
            long endTime = System.currentTimeMillis();
 
            // MLOPS: report predictions data: features, predictions, the 'null' arguments below
            // are for 'associationIds' (required for calculating accuracy - not included in this
            // snippet) and 'classNames' (no classes for regression deployment)
            mlops.reportPredictionsData(featureData, (List<?>)predictions, null, null);
 
            // MLOPS: report number of samples and time to produce predictions.
            mlops.reportDeploymentStats(samples.size(), endTime - startTime);
 
        } catch (Exception e) {
            System.err.println("Failed to report statistics, check the error below");
            e.printStackTrace();
            System.exit(-1);
        } finally {
            // MLOPS: shutdown
            if (mlops != null) {
                mlops.shutdown();
            }
        }
    }
 
    public static void main(String args[]) {
        String modelFilePath;
        String csvFilePath;
 
        try {
            modelFilePath = args[0];
            csvFilePath = args[1];
        } catch (Exception e) {
            String thisProgram = new java.io.File(MLOpsCodeGenRegressionExample.class.getProtectionDomain()
                    .getCodeSource()
                    .getLocation()
                    .getPath())
                    .getName();
            System.err.println("Usage: java -jar " + thisProgram + " <codeGenModelFilePath>" + " <csvFilePath>");
            System.exit(1);
            return;
        }
 
        MLOpsCodeGenRegressionExample codeGenExample = new MLOpsCodeGenRegressionExample(modelFilePath, csvFilePath);
        codeGenExample.run();
    }
}
 